### Data Analysis 3 / Assignment 1
#### Elene Zuroshvili
##### 2 February 2026

---

### Environment & Reproducibility

This project is fully reproducible.

All required Python dependencies are specified in the file `environment.yml`.
To reproduce the results on a new machine, please run these commands in your terminal:

```bash
conda env create -f environment.yml
conda activate airbnb-pricing
jupyter lab

---
### Part 1: Modeling
##### I chose to do the analysis with Copenhagen, Denmark & later added Stockholm, Sweden

In [62]:
# imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [63]:
PATH_CORE  = "data/listings_March_2024.csv"      # older quarter (10K+) copenhagen, March 2024
PATH_LATER = "data/listings_September_2025.csv"     # most recent Copenhagen, September 2025
PATH_OTHER = "data/listings_stockholm_september_2025.csv"  # Stockholm (3K+), September 2025

df_core  = pd.read_csv(PATH_CORE,  low_memory=False)
df_later = pd.read_csv(PATH_LATER, low_memory=False)
df_other = pd.read_csv(PATH_OTHER, low_memory=False)

print("core :", df_core.shape)
print("later:", df_later.shape)
print("other:", df_other.shape)

core : (21707, 75)
later: (22994, 79)
other: (4955, 79)


In [64]:
df_core.head()

id                         listing_url       scrape_id last_scraped  \
0  31094  https://www.airbnb.com/rooms/31094  20241230011627   2024-12-30   
1  32379  https://www.airbnb.com/rooms/32379  20241230011627   2024-12-30   
2  32841  https://www.airbnb.com/rooms/32841  20241230011627   2024-12-30   
3  38499  https://www.airbnb.com/rooms/38499  20241230011627   2024-12-30   
4  39055  https://www.airbnb.com/rooms/39055  20241230011627   2024-12-30   

            source                                               name  \
0  previous scrape  Beautiful, spacious, central, renovated Penthouse   
1  previous scrape  155 m2 artist flat on Vesterbro, with 2 bathrooms   
2  previous scrape                Cozy flat for Adults/Quiet for kids   
3      city scrape                     0 min. from everything in Cph.   
4      city scrape            Stylish apartment in central Copenhagen   

                                         description  \
0  Welcome to our home, we hope you will enjoy Wo...   
1  You enter a narrow entrance and feel the good ...   
2  Our flat is placed in a Central AND Quiet neig...   
3  It doesn't get more central than this. Histori...   
4  Big, bright, airy and attractive apartment dec...   

                               neighborhood_overview  \
0  What else is nearby?<br />To be honest, We thi...   
1  Værnedamsvej area is super hip area, we call i...   
2                                                NaN   
3  Area: <br />This is the eye of Cph. right betw...   
4  The flat is excellently located in Islands Bry...   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/miso/Hosting-...   129976  ...   
1  https://a0.muscache.com/pictures/miso/Hosting-...   140105  ...   
2  https://a0.muscache.com/pictures/28047599/8efd...   142143  ...   
3  https://a0.muscache.com/pictures/hosting/Hosti...   122489  ...   
4  https://a0.muscache.com/pictures/hosting/Hosti...   167511  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.82                   4.80                4.53   
1                        4.91                   4.89                4.70   
2                        5.00                   4.50                4.50   
3                        5.00                   4.93                4.89   
4                        4.94                   4.74                4.66   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              1   
1     NaN                f                              2   
2     NaN                f                              1   
3     NaN                f                              1   
4     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                           1   
2                                           1   
3                                           1   
4                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            1   
2                                            0   
3                                            0   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.11  
1                                           0              0.48  
2                                           0              0.04  
3                                           0              0.19  
4                                           0              0.64  

[5 rows x 75 columns]

In [65]:
df_later.head()

id                         listing_url       scrape_id last_scraped  \
0  23500  https://www.airbnb.com/rooms/23500  20250929042231   2025-09-29   
1  29118  https://www.airbnb.com/rooms/29118  20250929042231   2025-09-29   
2  31094  https://www.airbnb.com/rooms/31094  20250929042231   2025-09-30   
3  32379  https://www.airbnb.com/rooms/32379  20250929042231   2025-09-30   
4  32841  https://www.airbnb.com/rooms/32841  20250929042231   2025-09-29   

            source                                               name  \
0      city scrape                 Nice room - great area (Vesterbro)   
1      city scrape                    Best Location in Cool Istedgade   
2  previous scrape  Beautiful, spacious, central, renovated Penthouse   
3  previous scrape  155 m2 artist flat on Vesterbro, with 2 bathrooms   
4      city scrape                Cozy flat for Adults/Quiet for kids   

                                         description  \
0  Nice and cosy room in the heart of Copenhagen ...   
1                                                NaN   
2  Welcome to our home, we hope you will enjoy Wo...   
3  You enter a narrow entrance and feel the good ...   
4  Our flat is placed in a Central AND Quiet neig...   

                               neighborhood_overview  \
0                                                NaN   
1                                                NaN   
2  What else is nearby?<br />To be honest, We thi...   
3  Værnedamsvej area is super hip area, we call i...   
4                                                NaN   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/123766/fa4694...    92053  ...   
1  https://a0.muscache.com/pictures/b83c7003-367a...   125230  ...   
2  https://a0.muscache.com/pictures/miso/Hosting-...   129976  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...   140105  ...   
4  https://a0.muscache.com/pictures/28047599/8efd...   142143  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        5.00                   5.00                4.80   
1                        5.00                   4.84                4.74   
2                        4.82                   4.80                4.53   
3                        4.92                   4.89                4.71   
4                        5.00                   4.50                4.50   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              2   
1     NaN                f                              1   
2     NaN                f                              1   
3     NaN                f                              2   
4     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                           1   
2                                           1   
3                                           1   
4                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            0   
2                                            0   
3                                            1   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.08  
1                                           0              0.15  
2                                           0              0.10  
3                                           0              0.47  
4                                           0              0.04  

[5 rows x 79 columns]

In [66]:
df_other.head()

id                          listing_url       scrape_id last_scraped  \
0  164448  https://www.airbnb.com/rooms/164448  20250929042406   2025-09-29   
1  220851  https://www.airbnb.com/rooms/220851  20250929042406   2025-09-29   
2  238411  https://www.airbnb.com/rooms/238411  20250929042406   2025-09-29   
3  242188  https://www.airbnb.com/rooms/242188  20250929042406   2025-09-29   
4  273906  https://www.airbnb.com/rooms/273906  20250929042406   2025-09-29   

        source                                         name  \
0  city scrape  Double room in central Stockholm with Wi-Fi   
1  city scrape                      One room in appartement   
2  city scrape          Cozy apartment in central Stockholm   
3  city scrape  Single room in central Stockholm with Wi-Fi   
4  city scrape               Penthouse in central Stockholm   

                                         description  \
0  I am renting out a nice double room on the top...   
1                                           Welcome!   
2                                                NaN   
3  I am renting out a nice single room on the top...   
4                                                NaN   

                               neighborhood_overview  \
0                                                NaN   
1  Many restaurangs wery close and walkingdistanc...   
2  Restaurants, cafés, museums, art galleries, pa...   
3                                                NaN   
4                                                NaN   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/f56d8d10-a7fa...   784312  ...   
1  https://a0.muscache.com/pictures/2085606/7a706...   412283  ...   
2  https://a0.muscache.com/pictures/2806060/7fc68...  1250232  ...   
3  https://a0.muscache.com/pictures/2303148/55d1e...   784312  ...   
4  https://a0.muscache.com/pictures/2881091/f5404...  1432722  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.97                   4.84                4.76   
1                        4.88                   4.82                4.71   
2                        4.95                   4.85                4.64   
3                        4.97                   4.89                4.83   
4                        4.75                   5.00                5.00   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              2   
1     NaN                f                              2   
2     NaN                f                              1   
3     NaN                f                              2   
4     NaN                f                              2   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           1   
2                                           1   
3                                           0   
4                                           2   

  calculated_host_listings_count_private_rooms  \
0                                            2   
1                                            1   
2                                            0   
3                                            2   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              2.63  
1                                           0              0.39  
2                                           0              0.65  
3                                           0              2.68  
4                                           0              0.02  

[5 rows x 79 columns]

In [67]:
# checking the columns
print(df_core.columns.tolist()[:50])

['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name', 'description', 'neighborhood_overview', 'picture_url', 'host_id', 'host_url', 'host_name', 'host_since', 'host_location', 'host_about', 'host_response_time', 'host_response_rate', 'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood', 'host_listings_count', 'host_total_listings_count', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price', 'minimum_nights', 'maximum_nights', 'minimum_minimum_nights', 'maximum_minimum_nights', 'minimum_maximum_nights', 'maximum_maximum_nights', 'minimum_nights_avg_ntm', 'maximum_nights_avg_ntm', 'calendar_updated']


In [68]:
# Selecting core variables used for pricing
# These are stable, known at listing time, and economically meaningful

CORE_COLS = [
    "price",                           # target
    "room_type",                       # listing type
    "property_type",                   # property quality/type
    "accommodates",                    # capacity
    "bedrooms", "beds",                # size & comfort
    "bathrooms_text",                  # bathrooms info (more complete than bathrooms)
    "neighbourhood_cleansed",           # standardized location proxy
    "latitude", "longitude",            # spatial effects
    "host_is_superhost",                # trust / quality signal
    "host_listings_count",              # professional host proxy
    "minimum_nights", "maximum_nights", # booking constraints
    "amenities"                        # feature-engineered later
]

df_core = df_core[CORE_COLS].copy()


In [69]:
# Inspect missingness to guide imputation strategy
df_core.isna().mean().sort_values(ascending=False)

price                     0.425485
beds                      0.424886
bedrooms                  0.032800
host_is_superhost         0.008983
bathrooms_text            0.000461
host_listings_count       0.000092
room_type                 0.000000
property_type             0.000000
accommodates              0.000000
neighbourhood_cleansed    0.000000
latitude                  0.000000
longitude                 0.000000
minimum_nights            0.000000
maximum_nights            0.000000
amenities                 0.000000
dtype: float64

In [70]:
# Clean price: convert to numeric
# High missingness is expected (blocked / unavailable listings)

df_core["price"] = (
    df_core["price"]
    .str.replace(r"[\$,€]", "", regex=True)
    .str.replace(",", "", regex=False)
    .astype(float)
)

# Drop rows with missing price (cannot train without target)
df_core = df_core.dropna(subset=["price"])

# Log-transform price to reduce skew and outlier impact
df_core["log_price"] = np.log1p(df_core["price"])

In [71]:
df_core[["price", "log_price"]].describe()

price     log_price
count   12471.000000  12471.000000
mean     1390.213375      7.082073
std      1609.396411      0.513686
min        72.000000      4.290459
25%       895.000000      6.797940
50%      1150.000000      7.048386
75%      1600.000000      7.378384
max    100000.000000     11.512935

In [72]:
# Extract numeric bathrooms from text (e.g. "1 bath", "1.5 baths", "Shared bath")
# Bathrooms are a strong quality signal for pricing

df_core["bathrooms"] = (
    df_core["bathrooms_text"]
    .str.extract(r"(\d+\.?\d*)")
    .astype(float)
)

# Impute missing bathrooms with median
df_core["bathrooms"] = df_core["bathrooms"].fillna(df_core["bathrooms"].median())

In [73]:
# Beds and bedrooms represent size/comfort
# Missing values are imputed with median to avoid dropping listings

df_core["beds"] = df_core["beds"].fillna(df_core["beds"].median())
df_core["bedrooms"] = df_core["bedrooms"].fillna(df_core["bedrooms"].median())

In [74]:
# Superhost = quality / trust signal
# Map to binary; missing treated as non-superhost (conservative)

df_core["host_is_superhost"] = (
    df_core["host_is_superhost"]
    .map({"t": 1, "f": 0})
    .fillna(0)
)

# Host listings count = professional host proxy
df_core["host_listings_count"] = (
    df_core["host_listings_count"]
    .fillna(df_core["host_listings_count"].median())
)

In [75]:
# Categorical variables: missing -> "Unknown" to keep all rows

cat_cols = ["room_type", "property_type", "neighbourhood_cleansed"]

for c in cat_cols:
    df_core[c] = df_core[c].fillna("Unknown")

In [76]:
# Convert amenities text into binary features for most common amenities

from collections import Counter

amenities_series = (
    df_core["amenities"]
    .str.lower()
    .str.replace(r"[{}\"]", "", regex=True)
    .str.split(",")
)

amenity_counts = Counter(a.strip() for sub in amenities_series for a in sub)
top_amenities = [a for a, _ in amenity_counts.most_common(20)]

for amenity in top_amenities:
    df_core[f"amenity_{amenity.replace(' ', '_')}"] = amenities_series.apply(
        lambda x: int(amenity in [i.strip() for i in x])
    )

In [77]:
# Drop raw text columns after feature engineering

df_core = df_core.drop(columns=["bathrooms_text", "amenities"])

In [78]:
df_core.head()

price        room_type                property_type  accommodates  \
3  3000.0  Entire home/apt                 Entire condo             6   
4  2200.0  Entire home/apt                 Entire condo             4   
5   470.0     Private room        Private room in condo             2   
6   999.0  Entire home/apt                 Entire condo             4   
7   465.0     Private room  Private room in rental unit             1   

   bedrooms  beds neighbourhood_cleansed   latitude  longitude  \
3       2.0   2.0               Indre By  55.684288  12.573019   
4       2.0   2.0            Amager Vest  55.665070  12.583150   
5       1.0   1.0                Nrrebro  55.707100  12.553860   
6       2.0   0.0               Indre By  55.690630  12.570480   
7       1.0   1.0          Frederiksberg  55.684930  12.534340   

   host_is_superhost  ...  amenity_dishwasher  amenity_tv  amenity_hangers  \
3                1.0  ...                   1           0                1   
4                1.0  ...                   1           0                0   
5                1.0  ...                   0           0                0   
6                1.0  ...                   1           0                1   
7                1.0  ...                   0           0                1   

   amenity_hot_water_kettle  amenity_wine_glasses  amenity_dining_table  \
3                         0                     1                     1   
4                         1                     1                     1   
5                         0                     0                     0   
6                         1                     1                     1   
7                         0                     1                     1   

   amenity_oven  amenity_freezer  amenity_heating  amenity_dedicated_workspace  
3             1                1                1                            1  
4             1                1                1                            0  
5             0                0                0                            1  
6             1                1                1                            1  
7             0                1                0                            1  

[5 rows x 35 columns]

#### 2. Build 5 different predictive models (3p)

   a. OLS  

   b. LASSO  

   c. Random Forest  

   d. Boosting of your choice  

   e. Something else of your choice. Could be another boosting algo, or something else.

In [79]:
# Define target and features
y = df_core["log_price"]
X = df_core.drop(columns=["price", "log_price"])

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [80]:
# One-hot encode categorical variables
X_train = pd.get_dummies(X_train, drop_first=True)
X_test  = pd.get_dummies(X_test, drop_first=True)

# Align columns to avoid train/test mismatch
X_train, X_test = X_train.align(X_test, join="left", axis=1, fill_value=0)

In [81]:
# OLS baseline model

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np

ols = LinearRegression()
ols.fit(X_train, y_train)

y_pred_ols = ols.predict(X_test)

ols_r2 = r2_score(y_test, y_pred_ols)
ols_rmse = np.sqrt(mean_squared_error(y_test, y_pred_ols))

print("OLS R2:", round(ols_r2, 3))
print("OLS RMSE:", round(ols_rmse, 3))

OLS R2: 0.538
OLS RMSE: 0.355


In [82]:
# LASSO: linear model with feature selection via L1 penalty

from sklearn.linear_model import LassoCV

lasso = LassoCV(cv=5, random_state=42, n_jobs=-1)
lasso.fit(X_train, y_train)

y_pred_lasso = lasso.predict(X_test)

lasso_r2 = r2_score(y_test, y_pred_lasso)
lasso_rmse = np.sqrt(mean_squared_error(y_test, y_pred_lasso))

print("LASSO R2:", round(lasso_r2, 3))
print("LASSO RMSE:", round(lasso_rmse, 3))
print("Chosen alpha:", round(lasso.alpha_, 5))

LASSO R2: 0.471
LASSO RMSE: 0.379
Chosen alpha: 0.01289


In [83]:
# Number of non-zero coefficients (feature selection effect)
sum(lasso.coef_ != 0)

9

In [84]:
# Random Forest: nonlinear model capturing interactions

from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=200,
    max_depth=None,
    min_samples_leaf=5,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

rf_r2 = r2_score(y_test, y_pred_rf)
rf_rmse = np.sqrt(mean_squared_error(y_test, y_pred_rf))

print("RF R2:", round(rf_r2, 3))
print("RF RMSE:", round(rf_rmse, 3))

RF R2: 0.576
RF RMSE: 0.34


In [85]:
rf_importance = pd.Series(
    rf.feature_importances_,
    index=X_train.columns
).sort_values(ascending=False)

In [86]:
# Gradient Boosting: strong nonlinear learner for tabular data

from sklearn.ensemble import GradientBoostingRegressor

gbr = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

gbr.fit(X_train, y_train)

y_pred_gbr = gbr.predict(X_test)

gbr_r2 = r2_score(y_test, y_pred_gbr)
gbr_rmse = np.sqrt(mean_squared_error(y_test, y_pred_gbr))

print("GBR R2:", round(gbr_r2, 3))
print("GBR RMSE:", round(gbr_rmse, 3))

GBR R2: 0.573
GBR RMSE: 0.341


In [87]:
# Feature importance from GBR
gbr_importance = pd.Series(
    gbr.feature_importances_,
    index=X_train.columns
).sort_values(ascending=False)

In [88]:
# Elastic Net: combines LASSO + Ridge (handles correlated features)

from sklearn.linear_model import ElasticNetCV

enet = ElasticNetCV(
    l1_ratio=[0.1, 0.5, 0.9],
    cv=5,
    random_state=42,
    n_jobs=-1
)

enet.fit(X_train, y_train)

y_pred_enet = enet.predict(X_test)

enet_r2 = r2_score(y_test, y_pred_enet)
enet_rmse = np.sqrt(mean_squared_error(y_test, y_pred_enet))

print("Elastic Net R2:", round(enet_r2, 3))
print("Elastic Net RMSE:", round(enet_rmse, 3))

Elastic Net R2: 0.47
Elastic Net RMSE: 0.38


#### 3. Compare models in terms of fit and time. (3p)

* Create a horserace table. 

* Discuss performance. 

In [89]:
# Compare all models in a single table

results = pd.DataFrame({
    "Model": ["OLS", "LASSO", "Random Forest", "Gradient Boosting", "Elastic Net"],
    "R2": [ols_r2, lasso_r2, rf_r2, gbr_r2, enet_r2],
    "RMSE": [ols_rmse, lasso_rmse, rf_rmse, gbr_rmse, enet_rmse]
})

results.sort_values("RMSE")

Model        R2      RMSE
2      Random Forest  0.576424  0.339560
3  Gradient Boosting  0.573172  0.340861
0                OLS  0.537858  0.354681
1              LASSO  0.470964  0.379483
4        Elastic Net  0.469919  0.379858

The results show clear performance differences across model classes. The Random Forest achieves the best overall fit (R² ≈ 0.58, lowest RMSE), closely followed by Gradient Boosting, indicating that nonlinear models are better suited to capture the complex relationships between listing characteristics, location, and price. These models benefit from their ability to model interactions and nonlinear effects, which are common in housing and short-term rental markets.

The OLS model performs reasonably well as a baseline, explaining over half of the variance in log prices. This suggests that many price determinants have approximately linear effects. However, its performance is clearly dominated by the tree-based models, highlighting the limitations of purely linear assumptions.

Both LASSO and Elastic Net perform worse than OLS. While these models are useful for feature selection and handling multicollinearity, the regularization penalty appears to oversimplify the pricing relationship in this setting, leading to underfitting.

Overall, the results indicate that model flexibility matters for Airbnb pricing. Tree-based ensemble methods provide the best trade-off between fit and robustness, making them the preferred choice for operational pricing decisions.

#### 4. Take two models (RF and a boosting model) and analyze them. (3p)

* Show feature importance. 

* Compare the 10 most important features. 

* Discuss your findings. 

In [90]:
# Top 10 important features from Random Forest
rf_top10 = rf_importance.head(10)
rf_top10

bedrooms                           0.227049
accommodates                       0.176792
room_type_Private room             0.140601
longitude                          0.095615
latitude                           0.071129
neighbourhood_cleansed_Indre By    0.066141
minimum_nights                     0.032786
maximum_nights                     0.024820
host_listings_count                0.021402
bathrooms                          0.015975
dtype: float64

In [91]:
# Top 10 important features from Gradient Boosting
gbr_top10 = gbr_importance.head(10)
gbr_top10


accommodates                       0.253790
bedrooms                           0.240210
room_type_Private room             0.163697
longitude                          0.073868
neighbourhood_cleansed_Indre By    0.068527
latitude                           0.038253
minimum_nights                     0.029711
host_listings_count                0.027566
bathrooms                          0.025183
amenity_dishwasher                 0.011739
dtype: float64

In [92]:
# Side-by-side comparison of top features
pd.concat(
    [rf_top10.rename("RF"), gbr_top10.rename("GBR")],
    axis=1
)

RF       GBR
bedrooms                         0.227049  0.240210
accommodates                     0.176792  0.253790
room_type_Private room           0.140601  0.163697
longitude                        0.095615  0.073868
latitude                         0.071129  0.038253
neighbourhood_cleansed_Indre By  0.066141  0.068527
minimum_nights                   0.032786  0.029711
maximum_nights                   0.024820       NaN
host_listings_count              0.021402  0.027566
bathrooms                        0.015975  0.025183
amenity_dishwasher                    NaN  0.011739

Discussion of Feature Importance (RF vs. Gradient Boosting)

Both Random Forest and Gradient Boosting identify capacity and size-related variables as the dominant drivers of Airbnb prices. In both models, bedrooms and accommodates are the two most important features, indicating that larger listings with higher guest capacity command substantially higher prices. This aligns well with economic intuition and the pricing strategies of short-term rental markets.

The importance of room type (Private room) highlights a strong pricing discount relative to entire homes, reflecting segmentation between shared and fully private accommodations. This variable consistently ranks among the top features in both models, suggesting that categorical listing type is a key structural determinant of price.

Location effects are captured through both geographic coordinates and neighborhood indicators. Latitude and longitude appear as important continuous spatial variables, while the specific neighborhood indicator (Indre By, the city center) further confirms a strong central-location premium. The presence of both spatial and categorical location variables among the top features indicates that price gradients are smooth across space but also jump discretely at neighborhood boundaries.

Booking constraints such as minimum and maximum nights contribute moderately to price formation. Listings with stricter minimum stays tend to have higher nightly prices, consistent with hosts targeting longer-stay or higher-value guests.

Host characteristics play a smaller but non-negligible role. Host listings count appears in both models, suggesting that professional hosts price differently from casual hosts, potentially due to experience, portfolio strategies, or scale effects.

A notable difference between the models is that Gradient Boosting assigns slightly more weight to amenities (e.g., dishwasher), whereas Random Forest emphasizes broader structural and spatial features. This reflects boosting’s strength in capturing smaller, additive effects, while Random Forest focuses more on dominant, high-variance predictors.

Overall, both models tell a consistent story: size, capacity, room type, and location are the primary determinants of Airbnb prices, with amenities and host characteristics playing secondary but meaningful roles.

---

### Part II. Validity

5. Let us have two "live" datasets. (3p)

   A. Download another (later) date.  Repeat data wrangling

   B. Download another city in the same country or region, with at least 3K observations. Repeat data wrangling

6. Use your 5 core models on the A. Later date and B. Other city data. Compare model performance and discuss the experience. (3p)

7. 2 extra points can be given for something extra

In [93]:
# Reusable wrangling to ensure identical preprocessing across datasets ( this is the same as I used above for the first dataset)

def wrangle(df):
    df = df[CORE_COLS].copy()

    # price
    df["price"] = (
        df["price"].str.replace(r"[\$,€]", "", regex=True)
        .str.replace(",", "", regex=False)
        .astype(float)
    )
    df = df.dropna(subset=["price"])
    df["log_price"] = np.log1p(df["price"])

    # bathrooms
    df["bathrooms"] = (
        df["bathrooms_text"].str.extract(r"(\d+\.?\d*)").astype(float)
    )
    df["bathrooms"] = df["bathrooms"].fillna(df["bathrooms"].median())

    # beds / bedrooms
    df["beds"] = df["beds"].fillna(df["beds"].median())
    df["bedrooms"] = df["bedrooms"].fillna(df["bedrooms"].median())

    # superhost / host count
    df["host_is_superhost"] = df["host_is_superhost"].map({"t":1,"f":0}).fillna(0)
    df["host_listings_count"] = df["host_listings_count"].fillna(
        df["host_listings_count"].median()
    )

    # categoricals
    for c in ["room_type","property_type","neighbourhood_cleansed"]:
        df[c] = df[c].fillna("Unknown")

    # amenities
    amenities_series = (
        df["amenities"].str.lower()
        .str.replace(r"[{}\"]", "", regex=True)
        .str.split(",")
    )
    for amenity in top_amenities:
        df[f"amenity_{amenity.replace(' ', '_')}"] = amenities_series.apply(
            lambda x: int(amenity in [i.strip() for i in x])
        )

    return df.drop(columns=["bathrooms_text","amenities"])


In [94]:
df_later_w = wrangle(df_later)
df_other_w = wrangle(df_other)

print(df_later_w.shape, df_other_w.shape)

(13442, 35) (3190, 35)


In [95]:
# align features to training set
def prepare_Xy(df):
    y = df["log_price"]
    X = df.drop(columns=["price","log_price"])
    X = pd.get_dummies(X, drop_first=True)
    X = X.reindex(columns=X_train.columns, fill_value=0)
    return X, y

X_later, y_later = prepare_Xy(df_later_w)
X_other, y_other = prepare_Xy(df_other_w)

In [96]:
# evaluate all 5 models on later date
def eval_model(model, X, y):
    pred = model.predict(X)
    return (
        r2_score(y, pred),
        np.sqrt(mean_squared_error(y, pred))
    )

later_results = pd.DataFrame({
    "Model": ["OLS","LASSO","RF","GBR","ElasticNet"],
    "R2": [
        eval_model(ols, X_later, y_later)[0],
        eval_model(lasso, X_later, y_later)[0],
        eval_model(rf, X_later, y_later)[0],
        eval_model(gbr, X_later, y_later)[0],
        eval_model(enet, X_later, y_later)[0]
    ],
    "RMSE": [
        eval_model(ols, X_later, y_later)[1],
        eval_model(lasso, X_later, y_later)[1],
        eval_model(rf, X_later, y_later)[1],
        eval_model(gbr, X_later, y_later)[1],
        eval_model(enet, X_later, y_later)[1]
    ]
})
later_results

Model        R2      RMSE
0         OLS  0.491399  0.364356
1       LASSO  0.426292  0.386975
2          RF  0.584647  0.329266
3         GBR  0.538143  0.347209
4  ElasticNet  0.425590  0.387212

In [97]:
# evaluate all 5 models on other city
other_results = pd.DataFrame({
    "Model": ["OLS","LASSO","RF","GBR","ElasticNet"],
    "R2": [
        r2_score(y_other, m.predict(X_other))
        for m in [ols, lasso, rf, gbr, enet]
    ],
    "RMSE": [
        np.sqrt(mean_squared_error(y_other, m.predict(X_other)))
        for m in [ols, lasso, rf, gbr, enet]
    ]
})
other_results


Model          R2       RMSE
0         OLS -725.371058  19.590646
1       LASSO    0.354243   0.584123
2          RF    0.391368   0.567083
3         GBR    0.401794   0.562205
4  ElasticNet    0.354381   0.584060

In [98]:
# Unified performance table: Core vs Later vs Other (same 5 models, automatic)

core_results = results.copy()
core_results["Dataset"] = "Core (Copenhagen Mar 2024)"

later_r = later_results.copy()
later_r["Dataset"] = "Later (Copenhagen Sep 2025)"

other_r = other_results.copy()
other_r["Dataset"] = "Other city (Stockholm Sep 2025)"

all_results = pd.concat([core_results, later_r, other_r], ignore_index=True)

# Optional: consistent model names
name_map = {"RF": "Random Forest", "GBR": "Gradient Boosting", "ElasticNet": "Elastic Net"}
all_results["Model"] = all_results["Model"].replace(name_map)

# Make a compact wide table
perf_table = (
    all_results
    .pivot_table(index="Model", columns="Dataset", values=["R2", "RMSE"])
    .round(3)
)

perf_table

R2                              \
Dataset           Core (Copenhagen Mar 2024) Later (Copenhagen Sep 2025)   
Model                                                                      
Elastic Net                            0.470                       0.426   
Gradient Boosting                      0.573                       0.538   
LASSO                                  0.471                       0.426   
OLS                                    0.538                       0.491   
Random Forest                          0.576                       0.585   

                                                                        RMSE  \
Dataset           Other city (Stockholm Sep 2025) Core (Copenhagen Mar 2024)   
Model                                                                          
Elastic Net                                 0.354                      0.380   
Gradient Boosting                           0.402                      0.341   
LASSO                                       0.354                      0.379   
OLS                                      -725.371                      0.355   
Random Forest                               0.391                      0.340   

                                                                               
Dataset           Later (Copenhagen Sep 2025) Other city (Stockholm Sep 2025)  
Model                                                                          
Elastic Net                             0.387                           0.584  
Gradient Boosting                       0.347                           0.562  
LASSO                                   0.387                           0.584  
OLS                                     0.364                          19.591  
Random Forest                           0.329                           0.567

## Part II. Validity: Model Performance on New Data

### A. Later Date (Same City)

When applied to a later time period in the same city, model performance remains **stable and strong**, particularly for the nonlinear models. Random Forest achieves the best performance (R² ≈ 0.58), closely followed by Gradient Boosting. This indicates that the pricing relationships learned from the earlier quarter generalize well over time, despite potential seasonality, demand shifts, or supply changes.

Linear models (OLS, LASSO, Elastic Net) experience a modest drop in explanatory power compared to the training period. This suggests that while linear relationships remain relevant, they are less flexible in adapting to evolving market conditions. Overall, the results support **temporal validity** of the modeling approach, especially for tree-based methods.

---

### B. Other City (Cross-City Transfer)

Performance drops more noticeably when models are transferred to a different city, which is expected. Location-specific factors, market structure, and consumer preferences differ across cities, reducing direct transferability.

The OLS model performs extremely poorly, indicating strong **model misspecification** when linear assumptions and city-specific coefficients are applied outside the original market. In contrast, LASSO, Elastic Net, Random Forest, and Gradient Boosting all retain reasonable explanatory power (R² ≈ 0.35–0.40).

Among these, **Gradient Boosting performs best**, followed closely by Random Forest. This suggests that nonlinear models are better able to generalize structural pricing patterns (e.g., size, capacity, room type) across markets, even when absolute price levels differ.

---

### Overall Validity Assessment

* **Temporal validity**: Strong — models, especially Random Forest and Gradient Boosting, generalize well to later periods within the same city.
* **Geographic validity**: Moderate — performance declines across cities, but nonlinear models remain usable.
* **Key insight**: Pricing models should ideally be **retrained or fine-tuned per city**, but tree-based models provide the most robust starting point for cross-market deployment.
